src/ATokenVaultFactory.sol

```diff
@@ -0,0 +1,231 @@
+        address deployer,
+        VaultParams params
+    );
+
+    /*//////////////////////////////////////////////////////////////
+                            STATE VARIABLES
+    //////////////////////////////////////////////////////////////*/
+
+    /// @notice Array of all deployed vaults
+    address[] internal _allVaults;
+
+    /// @notice Mapping from underlying asset to deployed vaults
+    mapping(address => address[]) internal _vaultsByUnderlying;
+
+    /// @notice Mapping from deployer to deployed vaults
+    mapping(address => address[]) internal _vaultsByDeployer;
+
+    /// @notice Mapping to check if a vault was deployed by this factory
+    mapping(address => bool) internal _isVaultDeployed;
+
+    /// @notice Proxy admin address for all deployed vaults
+    address public immutable PROXY_ADMIN;
+
+    /*//////////////////////////////////////////////////////////////
+                                STRUCTS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Struct containing constructor parameters for vault deployment
+     */
+    struct VaultParams {
+        address underlying;
+        uint16 referralCode;
+        IPoolAddressesProvider poolAddressesProvider;
+        address owner;
+        uint256 initialFee;
+        string shareName;
+        string shareSymbol;
+        uint256 initialLockDeposit;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Constructor
+     * @param proxyAdmin The address that will be the admin of all deployed proxies
+     */
+    constructor(address proxyAdmin) {
+        require(proxyAdmin != address(0), "ZERO_ADDRESS_NOT_VALID");
+        PROXY_ADMIN = proxyAdmin;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                            EXTERNAL FUNCTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Deploy a new ATokenVault with the given parameters
+     * @param params All parameters needed for vault deployment and initialization
+     * @return vault The address of the deployed vault proxy
+     */
+    function deployVault(VaultParams memory params) public returns (address vault) {
+        require(params.underlying != address(0), "ZERO_ADDRESS_NOT_VALID");
+        require(address(params.poolAddressesProvider) != address(0), "ZERO_ADDRESS_NOT_VALID");
+        require(params.owner != address(0), "ZERO_ADDRESS_NOT_VALID");
+        require(params.initialLockDeposit > 0, "ZERO_INITIAL_LOCK_DEPOSIT");
+        require(bytes(params.shareName).length > 0, "EMPTY_SHARE_NAME");
+        require(bytes(params.shareSymbol).length > 0, "EMPTY_SHARE_SYMBOL");
+
+        // Transfer the initial lock deposit from caller to this contract
+        IERC20(params.underlying).transferFrom(
```

Comment:
let's use safeERC20 since underlying is arbitrary

Reply:
Done @ https://github.com/aave/Aave-Vault/pull/99/commits/3218db719ce9e92f99e5064cefadb123d66ea3e5